File: crates/spfs/spfs/cli/_cmd_reset.py
from typing import List
import sys
import argparse

import spfs

from colorama import Fore


def register(sub_parsers: argparse._SubParsersAction) -> None:

    reset_cmd = sub_parsers.add_parser("reset", help=_reset.__doc__)
    reset_cmd.add_argument(
        "--edit",
        "-e",
        action="store_true",
        help="mount the /spfs filesystem in edit mode (true if REF is empty or not given)",
    )
    reset_cmd.add_argument(
        "--ref",
        "-r",
        metavar="REF",
        help=(
            "The tag or id of the desired runtime, or the current runtime if not given."
            " Use '-' or an empty string to request an empty environment. Only valid"
            " if no paths are given"
        ),
    )
    reset_cmd.add_argument(
        "paths",
        metavar="PATH",
        nargs="*",
        help="Paths under /spfs to reset, or all paths if none given",
    )
    reset_cmd.set_defaults(func=_reset)


def _reset(args: argparse.Namespace) -> None:
    """Rebuild the current /spfs dir with the requested refs, removing any active changes."""

    config = spfs.get_config()
    repo = config.get_repository()
    runtime = spfs.active_runtime()
    if args.ref:

        if args.paths:
            print(
                f"{Fore.RED}Cannot specify both --ref and PATHs{Fore.RESET}",
                file=sys.stderr,
            )
            sys.exit(1)

        runtime.reset()
        runtime.reset_stack()
        if args.ref in ("-", ""):
            args.edit = True
        else:
            env_spec = spfs.tracking.EnvSpec(args.ref[0])
            for target in env_spec.tags:
                obj = repo.read_ref(target)
                runtime.push_digest(obj.digest())
    else:
        paths = _strip_spfs_prefix(args.paths)
        runtime.reset(*paths)

    if args.edit:
        runtime.set_editable(args.edit)

    spfs.remount_runtime(runtime)


def _strip_spfs_prefix(paths: List[str]) -> List[str]:
    out = []
    for path in paths:
        if path.startswith("/spfs"):
            path = path[len("/spfs") :]
        out.append(path)
    return out

File: crates/spfs/src/cli/cmd_reset.rs
// Copyright (c) 2021 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use clap::Args;
use spfs::prelude::*;

/// Reset changes, or rebuild the entire spfs directory
#[derive(Args, Debug)]
pub struct CmdReset {
    /// Mount the resulting runtime in edit mode
    ///
    /// Default to true if REF is empty or not given
    #[clap(short, long)]
    edit: bool,

    /// The tag or id to rebuild the runtime with.
    ///
    /// Uses current runtime stack if not given. Use '-' or
    /// an empty string to request an empty environment. Only valid
    /// if no paths are given
    #[clap(long = "ref", short)]
    reference: Option<String>,

    /// Glob patterns in the spfs dir of files to reset, defaults to everything
    paths: Vec<String>,
}

impl CmdReset {
    pub async fn run(&mut self, config: &spfs::Config) -> spfs::Result<i32> {
        let mut runtime = spfs::active_runtime().await?;
        let repo = config.get_local_repository().await?;
        if let Some(reference) = &self.reference {
            runtime.reset::<&str>(&[])?;
            runtime.status.stack.truncate(0);
            match reference.as_str() {
                "" | "-" => self.edit = true,
                _ => {
                    let env_spec = spfs::tracking::EnvSpec::parse(reference)?;
                    for target in env_spec.iter() {
                        let obj = repo.read_ref(target.to_string().as_ref()).await?;
                        runtime.push_digest(&obj.digest()?);
                    }
                }
            }
        } else {
            let paths = strip_spfs_prefix(&self.paths);
            runtime.reset(paths.as_slice())?;
        }

        if self.edit {
            runtime.status.editable = true;
        }

        runtime.save_state_to_storage().await?;
        spfs::remount_runtime(&runtime).await?;
        Ok(0)
    }
}

fn strip_spfs_prefix(paths: &[String]) -> Vec<String> {
    paths
        .iter()
        .map(|path| {
            path.strip_prefix("/spfs")
                .unwrap_or_else(|| path.as_ref())
                .to_owned()
        })
        .collect()
}
